app/services/migration/failure_manager.rb

```diff
@@ -0,0 +1,65 @@
+module Migration
+  class FailureManager
+    def record_failure(item, failure_message)
+      return unless data_migration && item.present?
+
+      @item = item
+      @failure_message = failure_message
+
+      write_failure
+      increment_counter
+    end
+
+    def all_failures
+      return unless data_migration
+
+      read_failures
+    end
+
+  private
+
+    attr_reader :data_migration, :item, :failure_message, :counter
+
+    def initialize(data_migration:)
+      @data_migration = data_migration
+      @counter = 1
+    end
+
+    def increment_counter
+      @counter += 1
+    end
+
+    def attributes
+      %i[
+        id
+        ecf_id
+      ].freeze
+    end
+
+    def failure
+      {
+        model: data_migration.model,
+        id: item.id,
+        failure_message:,
+      }
+    end
+
+    def migration_failure_key
+      @migration_failure_key ||= "migration_failure_#{data_migration.model}_#{data_migration.id}"
+    end
+
+    def write_failure
+      Rails.cache.write("#{migration_failure_key}_#{counter}", failure, expires_in: 1.month)
```

Comment:
I think storing the failures under a single key per data migration would simplify a lot of this (I don't think there Redis value size limit will be an issue). So having:

```
data_migration.id => {
  failure_message_1 => [item1, item2],
  failure_message_2 => [item1, item2]
}
```

Then we don't need to maintain a `counter` which could cause issues if we ended up with two instances of a `FailureManager` with the same `data_migration` recording failures. Having the failures grouped by failure message is probably how we would look at the data as well I imagine 🤔